.github/workflows/ci.yaml

```diff
@@ -0,0 +1,63 @@
+name: Continuous Integration
+
+on:
+  push:
+    branches:
+      - '**'
+      # - $default-branch
+  pull_request:
+    branches:
+      - '**'
+      # - $default-branch
+
+jobs:
+  build:
+    env:
+      COVER: true
+      GOPATH: ${{ github.workspace }}
+    runs-on: ubuntu-18.04
+    steps:
+
+    - name: Check out code into the Go module directory
+      uses: actions/checkout@v2
+      with:
+        path: ./src/github.com/${{ github.repository }}
+
+    - name: Set up Go 1.14
+      uses: actions/setup-go@v2
+      with:
+        go-version: ^1.14
+      id: go
+
+    - name: Get dependencies
+      run: |
+        cd src/github.com/${{ github.repository }}
+        curl -sfL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | sh -s -- -b $GOPATH/bin v1.24.0
+        GO111MODULE=on go mod download
+        curl -L https://codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-amd64 > ./cc-test-reporter
+        chmod +x ./cc-test-reporter
+
+    - name: Before Build
+      run: |
+        cd src/github.com/${{ github.repository }}
+        ./cc-test-reporter before-build
+
+    - name: Lint
+      run: |
+        cd src/github.com/${{ github.repository }}
+        make lint
+
+    - name: Build
+      run: |
+        cd src/github.com/${{ github.repository }}
+        make build
+
+    - name: Test
+      run: |
+        cd src/github.com/${{ github.repository }}
+        make test
```

Comment:
So I think this needs to be
```
./cc-test-reporter before-build
make test
./cc-test-reporter after-build --exit-code $? -t gocov
```
This should capture the exit code from the previous command, though I don't know if `make test` failing would cause this step to fail?

Reply:
If `make test` fails (non zero output), then next command won't run. I've included `./cc-test-reporter before-build` line [before](https://github.com/oauth2-proxy/oauth2-proxy/blob/5cd2e0e9eb1d132c8a98b31791deb84699d7b916/.github/workflows/ci.yaml#L43) `make lint`, `make build` is that incorrect?

Does `./cc-test-reporter after-build` script always needs `--exit-code` value?

> This needs to be in the test step, so I'd prefer if we had three separate jobs, which would give us three separate ticks right?

I didn't quite get it :D

Reply:
> If make test fails (non zero output), then next command won't run. I've included ./cc-test-reporter before-build line before make lint, make build is that incorrect?

It would be preferable if the script could continue on make exiting non-zero to still run the after build

The before and after is only needed for the test step, not for linting or building

> I didn't quite get it :D

Does it make sense now what we are aiming for?

Reply:
Making some progress. As you suggested, I'm running `cc-test-reporter` only before and after `make test` like this.
```
./cc-test-reporter before-build
make test
./cc-test-reporter after-build --exit-code $? -t gocov
```

However after build part is giving me following error. I even tried to manually giving `--exit-code 0`. How to resolve this?

```
./cc-test-reporter after-build --exit-code 0 -t gocov
ERRO[0000] could not find coverage file 
could not find any files in search paths for gocov. search paths were: , c.out 
Error: could not find any files in search paths for gocov. search paths were: , c.out
Usage:
  cc-test-reporter after-build [flags]

Flags:
  -s, --batch-size int               batch size for source files (default 500)
  -e, --coverage-endpoint string     endpoint to upload coverage information to (default "https://api.codeclimate.com/v1/test_reports")
  -t, --coverage-input-type string   type of input source to use [clover, cobertura, coverage.py, excoveralls, gcov, gocov, jacoco, lcov, simplecov, xccov]
      --exit-code int                exit code of the test run
  -r, --id string                    reporter identifier
      --insecure                     send coverage insecurely (without HTTPS)
  -p, --prefix string                the root directory where the coverage analysis was performed (default "/go/src/github.com/oauth2-proxy/oauth2-proxy")

Global Flags:
  -d, --debug   run in debug mode
```

Reply:
Alright check [this](https://github.com/shinebayar-g/oauth2-proxy/runs/1093635299?check_suite_focus=true). I think I got pretty close now. WDYT?

Reply:
Yeah that's looking pretty good, not entirely sure why the after build is complaining though, I assume the status code is remaining unset or something? 

On this PR it's having issues because the reporter ID isn't set, do I need to set this up as an actions environment variable or something?

Reply:
Status code is setting to `0` successfully. I've even printed that in [this](https://github.com/shinebayar-g/oauth2-proxy/runs/1141086075?check_suite_focus=true#step:7:2504) run.

For me even locally this line is not working properly. If `$TEST_STATUS` is equal to `0` below line should work right?

```
root@7030f51f7041:/go/src/github.com/oauth2-proxy/oauth2-proxy# ./cc-test-reporter after-build --exit-code 0 -t gocov
ERRO[0000] could not find coverage file 
could not find any files in search paths for gocov. search paths were: , c.out 
Error: could not find any files in search paths for gocov. search paths were: , c.out
Usage:
  cc-test-reporter after-build [flags]

Flags:
  -s, --batch-size int               batch size for source files (default 500)
  -e, --coverage-endpoint string     endpoint to upload coverage information to (default "https://api.codeclimate.com/v1/test_reports")
  -t, --coverage-input-type string   type of input source to use [clover, cobertura, coverage.py, excoveralls, gcov, gocov, jacoco, lcov, simplecov, xccov]
      --exit-code int                exit code of the test run
  -r, --id string                    reporter identifier
      --insecure                     send coverage insecurely (without HTTPS)
  -p, --prefix string                the root directory where the coverage analysis was performed (default "/go/src/github.com/oauth2-proxy/oauth2-proxy")

Global Flags:
  -d, --debug   run in debug mode
```

What is reported ID? Does it have to set by environment variable? Is that should be secret? Is that unique number for everyone?

Reply:
So we have a env for travis of `CC_TEST_REPORTER_ID=<our ID>`, however, it only sets this env var if the PR comes from a repository collaborator, so the after-build only actually works for PRs from collaborators. 

Code climate don't consider this to be a sensitive value, so we can put it into the action file if we wanted to, however, we need to ensure that forks don't run their builds and report to our code climate account.

Can we put into the action somehow a way so that it only runs on pushes to branches to this specific repo or PRs to this specific repo?

Reply:
I wonder if we can set it to an empty string when it's not set in the env already? I believe if the ENV var is set but empty it may just skip doing anything 🤔 

I've added it as a secret for actions for now, not sure that will have any effect though

Reply:
> Can we put into the action somehow a way so that it only runs on pushes to branches to this specific repo or PRs to this specific repo?

That's possible. Go to repo settings, go to secrets, add your secrets as K,V. We can expose that secrets into workflow jobs as Environment Variables. This secrets [are not exposed](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets#using-encrypted-secrets-in-a-workflow) to forks of the repo. So it should be safe.

Assuming you've already created secret named `CC_TEST_REPORTER_ID` this is not exposed to my fork as you can see [here](https://github.com/oauth2-proxy/oauth2-proxy/pull/750/checks?check_run_id=1170845144#step:7:2551).

I've created secret named `CC_TEST_REPORTER_ID` in my [fork settings](https://github.com/shinebayar-g/oauth2-proxy/runs/1170845084?check_suite_focus=true#step:7:2507) and it's some arbitrary number. Even though output is not appealing. Even it's wrong ID it should at least report that ID is wrong or something right? I think we're missing entirely other thing here.

Reply:
Perhaps for a first pass we need to set wrap the calls to the before and after build for CC in an if statement that checks if that variable is set or not? WDYT? That would allow us to separate this issue from the rest of the problem here and we can follow up later to work out why the CC isn't working

Reply:
Sorry, I didn't get that.

Reply:
As a temporary solution, in the test script, wrap each time we use the test reporter depending on if the variable is set.

```
if [ -z "$CC_TEST_REPORT_ID" ]; then
  ./cc-test-reporter ...
fi

```

Then we will be able to check if it's working or not based on PRs from the origin repo, then we can try expanding to PRs from external forks (which don't presently work anyway).

Why it's panicking when it's not set is really weird, I definitely don't think it used to do that

Reply:
Okay, it's done. Now at least it passes. Would you squash my commits?

Reply:
Results are looking good now. Will squash on merge yes

Did you want to add a changelog entry at all?

Reply:
Sure, how can I do that?

Reply:
Add an entry to https://github.com/oauth2-proxy/oauth2-proxy/blob/master/CHANGELOG.md#changes-since-v611 in the same way that others have done please

Reply:
Done, thank you.